Implement channel

jomuel opened the pull request.

jomuel committed:
feat: add channels (work in progress)

jomuel committed:
test: add a channel example

jomuel committed:
feat: channel implementation (wip)

jomuel committed:
feat: channel implementation (wip)

jomuel committed:
test: changes to channel example

spearman committed:
fix: fix endianness in channel envelope, fix msgpack in channel example

jomuel committed:
feat: finish channel implementation

jomuel committed:
docs: add doc comments to channel module

jomuel committed:
test: mock objects and unit tests for channel module

jomuel committed:
test: update channel examples

spearman committed:
docs: fix words in channel doc comments

spearman committed:
refactor: channel remove link getter and make link pub

jomuel committed:
refactor: unused imports and code in channel.rs

jomuel committed:
test: lint channel examples and unit tests

jomuel committed:
test: fix channel example (was stalled because of changed announce handling)

spearman committed:
test: channel_send integration test

jomuel committed:
lint: fix clippy warnings for tests and examples

jomuel committed:
fix: use correct validating key in link message proof

jomuel committed:
fix: check link message proofs for in-links too

jomuel committed:
lint: clean up channel client example

jomuel committed:
lint: channel server example

spearman commented on src/channel.rs:
```diff
@@ -0,0 +1,1217 @@
+            let result = self.incoming.send(message);
+
+            if let Err(e) = result {
+                log::error!(
```
Do we want to log an error here? The only error type SendError indicates "no active receivers". I think there could be a race between calling Channel::new and channel.subscribe() where the user could see an error.

The error display text is "channel closed", but really it just means there are currently no subscribers, but there could be later. My vote would be to change the log level of the message to debug. Another option would be to have Channel::new return (Channel, Receiver) so that subscribe() doesn't need to be manually called after initialization.

spearman reviewed (changes requested):
Looks good, one request to change log error statement and/or refactor to avoid likelihood of errors

jomuel committed:
feat: log warning for unhandled channel messages

jomuel committed:
feat: return a receiver from Channel::new()

there should always be at least one receiver for incoming messages,
so all messages are processed.

jomuel committed:
chore: adapt examples to changed Channel::new()

spearman committed:
Merge branch 'main' into channel

spearman reviewed (approved):
Looks good
